pxf/pxf-ignite/src/main/java/org/apache/hawq/pxf/plugins/ignite/IgniteAccessor.java

```diff
@@ -0,0 +1,502 @@
+    /**
+     * writeNextObject() implementation
+     */
+    @Override
+    public boolean writeNextObject(OneRow currentRow) throws Exception {
+        boolean currentRowInBuffer = bufferWrite.add(currentRow);
+
+        if (!isWriteActive) {
+            if (!currentRowInBuffer) {
+                LOG.error("writeNextObject(): Failed (not enough memory in 'bufferWrite')");
+                throw new OutOfMemoryError("writeNextObject(): not enough memory in 'bufferWrite'");
+            }
+            LOG.info("Ignite write request. Query: '" + queryWrite + "'");
+            sendInsertRestRequest(queryWrite);
+            bufferWrite.removeFirst();
+            isWriteActive = true;
+            return true;
+        }
+
+        if ((bufferWrite.size() >= bufferSize) || (!currentRowInBuffer)) {
+            sendInsertRestRequest(queryWrite);
+        }
+
+        if (!currentRowInBuffer) {
+            if (!bufferWrite.add(currentRow)) {
+                LOG.error("writeNextObject(): Failed (not enough memory in 'bufferSend')");
+                throw new OutOfMemoryError("writeNextObject(): not enough memory in 'bufferSend'");
+            }
+        }
+        
+        return true;
+    }
+
+    /**
+     * closeForWrite() implementation
+     */
+    @Override
+    public void closeForWrite() throws Exception {
+        if (!bufferWrite.isEmpty()) {
+            sendInsertRestRequest(queryWrite);
+        }
+        if (isWriteActive) {
+            // At this point, the request must have finished successfully; otherwise an exception would be thrown before
+            LOG.info("Ignite write request finished successfully. Query: '" + queryWrite + "'");
+        }
+        isWriteActive = false;
+    }
+
+    /**
+     * Build HTTP GET query for Ignite REST API with command 'qryfldexe'
+     * 
+     * @param querySql SQL query with filter constants. The constants must replaced by "?" if 'filterConstants' is not given
+     * @param filterConstants
+     * 
+     * @return Prepared HTTP query. The query will be properly encoded with {@link java.net.URLEncoder}
+     * 
+     * @throws UnsupportedEncodingException from {@link java.net.URLEncoder.encode()}
+     */
+    private String buildQueryFldexe(String querySql, ArrayList<String> filterConstants) throws UnsupportedEncodingException {
+        StringBuilder sb = new StringBuilder();
+        sb.append("http://");
+        sb.append(igniteHost);
+        sb.append("/ignite");
+        sb.append("?");
+        sb.append("cmd=qryfldexe");
+        sb.append("&");
+        sb.append("pageSize=0");
+        sb.append("&");
+        if (cacheName != null) {
+            sb.append("cacheName=");
+            // Note that Ignite supports only "good" cache names (those that should be left unchanged by the URLEncoder.encode())
+            sb.append(URLEncoder.encode(cacheName, "UTF-8"));
+            sb.append("&");
+        }
+        int counter = 1;
+        if (filterConstants != null) {

```

Comment:
could you please the filter construction? it looks like it creates something like
`arg1=A&arg2=B&arg3=C`
how do we know what field argX stands for? and if we want to filter by the last field, how can it be done? (I would think in that case we will have to pass the filterConstants with empty values as placeholders for the other fields?)
If I understand correctly, this code is never executed because we always send filterConstants as null, so I just want to make sure it will work as expected.

Reply:
This behaviour is correct.

Apache Ignite [supports](https://apacheignite.readme.io/docs/rest-api#section-sql-fields-query-execute) passing the constraints' constants separately from the constraint expression itself. I am not sure why there is such an option; maybe it is there for safety (to protect Ignite from SQL injections, if the request is formed by, say, javascript on some webpage).

This feature may be implemented in the future, so I left the code as it is. To implement the feature, relatively serious changes must be done both in `IgnitePartitionFragmenter` and `WhereSqlBuilder`; however, the effect of those changes is not evident.